Implement persistence with the new structures

evanlinjin opened the pull request:
### Description

This is part of #895 and #971

* Introduce a more generic version of the `keychain::persist::*` structures that only needs a single generic for the changeset type.

Additional changes:

* The `Append` trait has a new method `is_empty`.
* Introduce `Store` structure for `bdk_file_store` (which implements `PersistBackend`).

### Changelog notice

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

LLFourn commented on crates/file_store/src/store.rs:
```diff
@@ -0,0 +1,324 @@
+    /// directly after the appended changeset.
+    ///
+    /// **WARNING**: This method does not detect whether the changeset is empty or not, and will
+    /// append an empty changeset to the file (not catastrophic, just a waste of space).
```
I think it should do this check. I'd add `is_empty` onto `Append`.

evanlinjin replied:
Added in d2e7b48e2c76a34f065705053f6e6955d6878d22

bodymindarts commented on crates/chain/src/persist.rs:
```diff
@@ -0,0 +1,89 @@
+/// A persistence backend for [`Persist`].
+///
+/// `T` represents the tracker, the in-memory data structure which we wish to persist.
+pub trait PersistBackend<T, C> {
```
Has any consideration been given to making this an [`#[async_trait]`](https://docs.rs/async-trait/latest/async_trait/)? IO operations like write / load should generally not be executed synchronously - especially if the persistence backend is a networked database.

LLFourn replied:
This is a good point. Some background on this trait. We'd rather this trait not exist but in order to preserve `Wallet`'s (over) simple API that automatically persists every change you make to it we had to add this trait. I think that if you are using a networked backend you are doing something sufficiently sophisticated to justify not using `bdk::Wallet` (and not using this trait) which is more meant as an example starting point kind of thing.

@evanlinjin thoughts?

evanlinjin replied:
Yes.

In general, the structures in `bdk_chain` have a simple API and easy to design persistence around without requiring a trait. Methods that make changes return a "changeset" which the caller can persist.

evanlinjin replied:
Maybe if we want to do an async version of `Wallet`, then it may be a good idea to have an async trait.

bodymindarts replied:
I see so for more sophisticated use cases you suggest using `bdk_chain` directly. Thanks for this insight!

evanlinjin committed:
[persist_redesign] Introduce redesigned `persist` types

This is a more generic version of `keychain::persist::*` structures.

Additional changes:

* The `Append` trait has a new method `is_empty`.
* Introduce `Store` structure for `bdk_file_store`.

LLFourn commented on crates/file_store/src/store.rs:
```diff
@@ -0,0 +1,289 @@
+    /// The file must have been opened with read and write permissions.
+    ///
+    /// [`File`]: std::fs::File
+    pub fn new(magic: &'a [u8], mut db_file: File) -> Result<Self, FileError> {
```
there is no docs for this very *magical* function argument.

LLFourn commented on crates/file_store/src/store.rs:
```diff
@@ -0,0 +1,289 @@
+    pub fn new(magic: &'a [u8], mut db_file: File) -> Result<Self, FileError> {
+        db_file.rewind()?;
+
+        let mut magic_buf = Vec::from_iter((0..).take(magic.len()));
```
```suggestion
        let mut magic_buf = vec![0u8;magic.len];
```

LLFourn commented on crates/file_store/src/entry_iter.rs:
```diff
@@ -0,0 +1,100 @@
+{
+    type Item = Result<T, IterError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
```
The logic of `EntryIter` seems to have gotten more complex without much benefit. I think it was better to let the caller seek to the right position before creating the `EntryIter` but perhaps I am missing something.

evanlinjin replied:
I think the changes result in a simpler API. Before, when we construct an `EntryIter` we need to return a `Result` (because seek may error), and the `EntryIter` itself also iterates over `Result`s.

LLFourn commented on crates/file_store/src/store.rs:
```diff
@@ -0,0 +1,289 @@
+        serde::Serialize,
+        serde::Deserialize,
+    )]
+    enum TestKeychain {
```
this is unused?

evanlinjin replied:
Unfortunately.

LLFourn commented on crates/file_store/src/store.rs:
```diff
@@ -0,0 +1,289 @@
+
+    #[derive(Debug, Default, serde::Serialize, serde::Deserialize)]
+    struct TestChangeSet {
+        pub changes: Vec<String>,
```
:bulb: : just impl `Append` for `Vec<T>` and use it directly in the test.

It could be cool to also implement `Append` for any `(A, B) where A: Append, B: Append` etc so you can easily add your "updated to" kind of variable for block syncing. Can be done at another time.

LLFourn commented on crates/chain/src/tx_data_traits.rs:
```diff
@@ -64,20 +64,35 @@ impl<A: Anchor> Anchor for &'static A {
     fn append(&mut self, other: Self);
+
+    /// Returns whether the structure is considered empty.
+    fn is_empty(&self) -> bool;
```
Let's say I wanted to implement `Append` for a `u32` to represent a monotonically increasing value -- I guess `u32::MIN` would return true for `is_empty`.

evanlinjin replied:
I think implementing `Append` for `u32` is a rather confusing combination. `Option<u32>` or using a wrapped type makes more sense for me.

LLFourn reviewed (approved):
I left a few nits. Otherwise LGTM.

evanlinjin committed:
Add more `impl`s for `Append` and docs for file store `magic`

danielabrozzoni commented on crates/chain/src/persist.rs:
```diff
@@ -0,0 +1,89 @@
+#[derive(Debug)]
+pub struct Persist<B, C> {
+    backend: B,
+    stage: C,
```
nit: maybe it should be S for stage, instead of C?

evanlinjin replied:
`C` for changeset though :open_mouth:

danielabrozzoni replied:
Then do `changeset: C` :P

notmandatory replied:
I support `changeset: C` 🙂

danielabrozzoni commented on crates/chain/src/persist.rs:
```diff
@@ -0,0 +1,89 @@
+    ///
+    /// Returns a backend-defined error if this fails.
+    pub fn commit(&mut self) -> Result<(), B::WriteError> {
+        let mut temp = C::default();
+        core::mem::swap(&mut temp, &mut self.stage);
+        self.backend.write_changes(&temp)
```
Mhhhh maybe put a small comment here, it took me a second to figure out what you were trying to do (commit changes and clean self.stage at the same time)

Also, a shorter way of writing this, but not necessarily a cleaner one, might be:
```
self.backend.write_changes(core::mem::take(&mut self.stage))
```

LLFourn replied:
actually shouldn't we only clear it in the case that it succeeds.

evanlinjin replied:
That's a good point

evanlinjin replied:
```rust
    /// Commit the staged changes to the underlying persistance backend.
    ///
    /// Changes that are committed (if any) are returned.
    ///
    /// # Error
    ///
    /// Returns a backend-defined error if this fails.
    pub fn commit(&mut self) -> Result<Option<C>, B::WriteError> {
        if self.stage.is_empty() {
            return Ok(None);
        }
        self.backend
            .write_changes(&self.stage)
            // if written successfully, take and return `self.stage`
            .map(|_| Some(core::mem::take(&mut self.stage)))
    }
```

evanlinjin replied:
96b10751325dc1bced09f9e864438216400e4ba3

danielabrozzoni commented on crates/file_store/src/entry_iter.rs:
```diff
@@ -0,0 +1,100 @@
+    db_file: Option<&'t mut File>,
+
+    /// The file position for the first read of `db_file`.
+    start_pos: Option<u64>,
```
nit: maybe expand in the docs why it's an Option and what happens when this is None
